crates/project/src/environment.rs

```diff
@@ -169,53 +175,46 @@ impl From<EnvironmentOrigin> for String {
 async fn load_shell_environment(
     _dir: &Path,
     _load_direnv: &DirenvSettings,
-) -> Result<HashMap<String, String>> {
-    Ok([("ZED_FAKE_TEST_ENV".into(), "true".into())]
+) -> Result<(HashMap<String, String>, Option<String>)> {
+    let fake_env = [("ZED_FAKE_TEST_ENV".into(), "true".into())]
         .into_iter()
-        .collect())
+        .collect();
+    Ok((fake_env, None))
 }
 
 #[cfg(not(any(test, feature = "test-support")))]
 async fn load_shell_environment(
     dir: &Path,
     load_direnv: &DirenvSettings,
-) -> Result<HashMap<String, String>> {
+) -> Result<(HashMap<String, String>, Option<String>)> {
+    use crate::direnv::{DirenvLoadError, DirenvWarning};
     use anyhow::{anyhow, Context};
     use std::path::PathBuf;
     use util::parse_env_output;
 
-    async fn load_direnv_environment(dir: &Path) -> Result<Option<HashMap<String, String>>> {
-        let Ok(direnv_path) = which::which("direnv") else {
-            return Ok(None);
-        };
-
-        let direnv_output = smol::process::Command::new(direnv_path)
-            .args(["export", "json"])
-            .current_dir(dir)
-            .output()
-            .await
-            .context("failed to spawn direnv to get local environment variables")?;
-
-        anyhow::ensure!(
-            direnv_output.status.success(),
-            "direnv exited with error {:?}. Stderr:\n{}",
-            direnv_output.status,
-            String::from_utf8_lossy(&direnv_output.stderr)
-        );
-
-        let output = String::from_utf8_lossy(&direnv_output.stdout);
-        if output.is_empty() {
-            return Ok(None);
-        }
-
-        Ok(Some(
-            serde_json::from_str(&output).context("failed to parse direnv output")?,
-        ))
-    }
-
+    let mut warning = DirenvWarning::default();
     let direnv_environment = match load_direnv {
         DirenvSettings::ShellHook => None,
-        DirenvSettings::Direct => load_direnv_environment(dir).await.log_err().flatten(),
+        DirenvSettings::Direct => match crate::direnv::load_direnv_environment(dir).await {
+            Ok(env) => env,
+            Err(err) => {
+                match err {
+                    DirenvLoadError::Io(io_error) => {
+                        log::error!("failed to spawn direnv with error: {}", io_error);
+                    }
+                    DirenvLoadError::NonZeroResult { status, stderr } => {
+                        warning = DirenvWarning::from(format!(
+                            "Loading direnv environment failed (exit code {status})\nStderr:\n{}",
+                            String::from_utf8_lossy(&stderr)
+                        ));
+                    }
+                    DirenvLoadError::Serde(serde_error) => {
+                        log::error!("direnv returned incorrect json: {serde_error}");
+                    }
+                }
+                None
+            }
+        },
```

Comment:
Why not make `load_direnv_environment` return `Option<...>`? That way you can keep all the logic inside the `direnv` file, you don't have to expose the `DirenvLoadError`.

Reply:
Makes sense